### PR TITLE
item 6150:  Fb prioritized updates domain designer test

### PR DESCRIPTION
**Pull request by labkey-chrisj**
this includes some synchronizing for more robust behavior on newer chromedriver
this adds testing for the prioritized update behavior
- user is prompted to confirm when navigating away with dirty form
- name field expands field row
this does not automate drag-and-drop re-ordering

### CODE REVIEW

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 1]
```diff
@@ -891,19 +893,122 @@ public void testLookUpFieldList() throws IOException, CommandException
...
+
+        DomainDesignerPage domainDesignerPage = DomainDesignerPage.beginAt(this, getProjectName(), "exp.materials", sampleSetName);
+        DomainFormPanel domainFormPanel = domainDesignerPage.fieldProperties(sampleSetName);
 
+        // now attempt to mark 'manufacturer' field 'required'
+        domainFormPanel.getField("manufacturer")
+                .setRequiredField(true);
+        domainDesignerPage.clickSave();     // expect error warning here; this should warn the user
+        String expectedWarning = domainDesignerPage.waitForAnyAlert();
+        assertTrue(expectedWarning.contains("cannot be required when it contains rows with blank values."));
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: Do we want a test case that verifies you **can** mark an existing field as required if there aren't any rows with blank values?

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 2]
```diff
@@ -891,19 +893,122 @@ public void testLookUpFieldList() throws IOException, CommandException
...
+        getDriver().switchTo().alert().dismiss();
+        // ensure current location
+        assertEquals(currentUrl, getDriver().getCurrentUrl());
+
+        domainDesignerPage = new DomainDesignerPage(getDriver()); // we've been refreshing; things are possibly stale
+        domainDesignerPage.clickCancel().saveChanges();
+    }
+
+    @Test
+    public void verifyWarningForMarkingExistingSampleSetFieldRequired() throws Exception
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: Is this special sample set behavior? If we're just using sample sets to verify universal behavior, having `SampleSet` in the name is misleading.

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 1 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: That's not a bad idea.  I'll add it

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 2 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: I'm not sure; my understanding was that this needed to work specifically on a sampleset.  It's definitely worth following up to see if this is/should be universal/true for every domain kind

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 3]
```diff
@@ -891,19 +893,122 @@ public void testLookUpFieldList() throws IOException, CommandException
...
+        // now navigate away
+        getDriver().navigate().to(homeUrl);
+        // dismiss the alert
+        shortWait().until(ExpectedConditions.alertIsPresent());
+        getDriver().switchTo().alert().dismiss();
+        // ensure current location
+        assertEquals(currentUrl, getDriver().getCurrentUrl());
+
+        domainDesignerPage = new DomainDesignerPage(getDriver()); // we've been refreshing; things are possibly stale
+        domainDesignerPage.clickCancel().saveChanges();
```
[the remark this answers is not in the document]
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: good call.

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 4]
```diff
@@ -891,19 +893,122 @@ public void testLookUpFieldList() throws IOException, CommandException
...
+
+        // now navigate away
+        getDriver().navigate().to(homeUrl);
+        // dismiss the alert
+        shortWait().until(ExpectedConditions.alertIsPresent());
+        getDriver().switchTo().alert().dismiss();
+        // ensure current location
+        assertEquals(currentUrl, getDriver().getCurrentUrl());
+
+        domainDesignerPage = new DomainDesignerPage(getDriver()); // we've been refreshing; things are possibly stale
```
[the remark this answers is not in the document]
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: fair enough.  tested to see if it was true, and yes... it was

File: src/org/labkey/test/components/domain/UnsavedChangesModalDialog.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    protected class ElementCache extends ModalDialog.ElementCache
+    {
+//        final WebElement yesSaveChangesBtn = Locator.css("button").withText("Yes, Save Changes")
+//                .findWhenNeeded(this);
+//
+//        final WebElement noDiscardChangesBtn = Locator.css("button").withText("No, Discard Changes")
+//                .findWhenNeeded(this);
+    }
```
[the remark this answers is not in the document]
Comment (labkey-chrisj) [reply src/org/labkey/test/components/domain/UnsavedChangesModalDialog.java]: done

**Pull request merged**